### ISSUE TITLE
Memory leak after at the end of an epoch of training with OpenCLIP on a WebDataset

### ISSUE THREAD

**Issue by fartashf** [webdataset/webdataset#365]
I observe a potential memory leak when training with OpenCLIP on a WebDataset. The memory usage is flat during an epoch and then increases at the beginning of the next epoch (evaluation disabled). If I set `persistent_workers=False` in the following line, the memory leak goes away.
https://github.com/mlfoundations/open_clip/blob/58e4e39aaabc6040839b0d2a7e8bf20979e4558a/src/training/data.py#L422

This screenshot shows a mock short training that if trained longer leads to RAM OOM for `persistent_workers=True`.
![W B Chart 6_15_2024, 8_48_40 PM](https://github.com/webdataset/webdataset/assets/7307513/c8d80242-c489-474e-961d-bcab86f3f448)

Setting `persistent_workers=False` does not seem to be a practical solution as per FAQ:
https://github.com/webdataset/webdataset/blob/5b12e0ba78bfb64741add2533c5d1e4cf088ffff/FAQ.md?plain=1#L1401-L1408
I'm wondering if this is related to how OpenCLIP creates a webdataset pipeline.
The OpenCLIP/WebDataset integration design was discussed in the following issue and alternative implementations were suggested:
https://github.com/webdataset/webdataset/issues/169
My guess is the shards that have partially been processed in an epoch will not be properly cleaned up from memory after the epoch is reset.
Another potential related issue but in OpenCLIP there is no explicit itertools.cycle: https://github.com/webdataset/webdataset/issues/357

**Comment by tmbdev** [comment]
It would be useful to figure out where that memory leak comes from.

Can you look with a memory tracing tool to see what objects are increasing in number?

It would also be worth figuring out whether this disappears if you eliminate shuffle buffers, since that is really the only location where WebDataset itself retains memory.

**Comment by fartashf** [comment]
Without `wds.shuffle`, the memory leak is still the same. Any memory tracing tool you'd recommend?

**Comment by fartashf** [comment]
I made a small script that reproduces the leak. I tried `tracemalloc` and `memray` but they don't show what's taking up the memory. One can monitor the memory usage going up running `watch -n1 free -g`.

```python
import logging                                                                                                                                                                                                [28/199]
import math
from tqdm import tqdm

import torch
import webdataset as wds


def filter_no_caption_or_no_image(sample):
    has_caption = ('txt' in sample)
    has_image = ('png' in sample or 'jpg' in sample or 'jpeg' in sample or 'webp' in sample)
    return has_caption and has_image


def log_and_continue(exn):
    """Call in an exception handler to ignore any exception, issue a warning, and continue."""
    logging.warning(f'Handling webdataset error ({repr(exn)}). Ignoring.')
    return True


def get_data(input_shards, batch_size, world_size, workers, num_samples,
             epoch=0, floor=False, no_persistent_workers=False):
    pipeline = [wds.SimpleShardList(input_shards)]
    pipeline.extend([
        wds.tarfile_to_samples(handler=log_and_continue),
        # wds.shuffle(
        #     bufsize=_SAMPLE_SHUFFLE_SIZE,
        #     initial=_SAMPLE_SHUFFLE_INITIAL,
        # ),
    ])
    pipeline.extend([
        wds.select(filter_no_caption_or_no_image),
        wds.decode("pilrgba", handler=log_and_continue),
        wds.rename(image="jpg;png;jpeg;webp", text="txt"),
        wds.map_dict(image=lambda x: torch.zeros((3, 224, 224)), text=lambda text:
                     torch.zeros((77,))),
        wds.to_tuple("image", "text"),
        wds.batched(batch_size, partial=False)
    ])
    dataset = wds.DataPipeline(*pipeline)
    # roll over and repeat a few samples to get same number of full batches on each node
    round_fn = math.floor if floor else math.ceil
    global_batch_size = batch_size * world_size
    num_batches = round_fn(num_samples / global_batch_size)
    num_workers = max(1, workers)
    num_worker_batches = round_fn(num_batches / num_workers)  # per dataloader worker
    num_batches = num_worker_batches * num_workers
    num_samples = num_batches * global_batch_size
    dataset = dataset.with_epoch(num_worker_batches)  # each worker is iterating over this
    dataloader = wds.WebLoader(
        dataset,
        batch_size=None,
        shuffle=False,
        num_workers=workers,
        persistent_workers=(workers > 0 and not no_persistent_workers),
    )
    dataloader.num_batches = num_batches
    dataloader.num_samples = num_samples
    return dataloader


if __name__ == '__main__':
    input_shards = "pipe:aws s3 cp s3://some-bucket/some-dataset/{00000000..00001023}.tar -"
    batch_size = 128
    world_size = 1
    workers = 4
    num_samples = 1024  # 2**20
    no_persistent_workers = False
    dataloader = get_data(input_shards, batch_size, world_size, workers, num_samples,
                          no_persistent_workers=no_persistent_workers)
    for epoch in range(100):
        for d in tqdm(dataloader, total=num_samples//batch_size):
            pass
```

**Comment by fartashf** [comment]
I investigated this a bit more. There are two issues here: 1) a pipe opened using `gopen` does not seem to ever close. 2) there seems to be no persistent workers logic.

Issue 1: The leak in the above code and similarly in [OpenCLIP](https://github.com/mlfoundations/open_clip/blob/fc5a37b72d705f760ebbc7915b84729816ed471f/src/open_clip_train/data.py#L328) happens because in each epoch a new iterator is created that calls `url_opener` but at the end of the epoch the stream opened by [`gopen.gopen`](https://github.com/webdataset/webdataset/blob/5b12e0ba78bfb64741add2533c5d1e4cf088ffff/webdataset/tariterators.py#L92) is never closed. The memory usage only increases when one uses `pipe:aws cp -` but doesn't happen with `pipe:cat`. So this could be a combination of how the `aws cp -` command is handling an unclosed pipe because calling `gc.collect()` is not helping.

Issue 2: My understanding is the implementation of `__iter__` by [webdataset](https://github.com/webdataset/webdataset/blob/5b12e0ba78bfb64741add2533c5d1e4cf088ffff/webdataset/pipeline.py#L78) does not have the logic of persistent workers from PyTorch's [`DataLoader.__iter__`](https://github.com/pytorch/pytorch/blob/fa5f5727484d78aa2ed8668c11550431bd415f5c/torch/utils/data/dataloader.py#L471) as it is inheriting from `IterableDataset` instead of `DataLoader`. Maybe this was by design and people are supposed to wrap a wds dataloader by a PyTorch dataloader? That doesn't seem right to me. But if the inheritance needs to change, one probably needs to inherit `__iter__` directly and then implement `_get_iterator` and potentially `_reset` methods.

**Comment by fartashf** [comment]
The simplest workaround in OpenCLIP is to reduce the number of epochs so that the dataloader doesn't reset too frequently. Specifically for OpenCLIP, adjust `--train-num-samples` (number of samples in dataset) and `--epochs` so that the total number of seen samples stays the same but divided into fewer epochs. The downside is the number of evaluations and saved checkpoints will be limited. In my case going from 100 epochs to 10 was good enough.
This is not a solution in any way and not necessarily applicable to code other than OpenCLIP.
